src/client/tile.cpp

```diff
@@ -775,205 +770,79 @@ void TextureSource::rebuildImagesAndTextures()
-				screen.Width /2 + (screen.Width  * (scaling / 2)),
-				screen.Height/2 + (screen.Height * (scaling / 2))
-			);
-
-		glReadPixels(source.UpperLeftCorner.X, source.UpperLeftCorner.Y,
-				partsize.Width, partsize.Height, GL_RGBA,
-				GL_UNSIGNED_BYTE, pixels);
-
-		driver->endScene();
-
-		// Drop scene manager
-		smgr->drop();
-
-		unsigned int pixelcount = partsize.Width*partsize.Height;
-
-		u8* runptr = pixels;
-		for (unsigned int i=0; i < pixelcount; i++) {
-
-			u8 B = *runptr;
-			u8 G = *(runptr+1);
-			u8 R = *(runptr+2);
-			u8 A = *(runptr+3);
-
-			//BGRA -> RGBA
-			*runptr = R;
-			runptr ++;
-			*runptr = G;
-			runptr ++;
-			*runptr = B;
-			runptr ++;
-			*runptr = A;
-			runptr ++;
-		}
-
-		video::IImage* inventory_image =
-				driver->createImage(irr::video::ECF_A8R8G8B8, params.dim);
-
-		rawImage->copyToScaling(inventory_image);
-		rawImage->drop();
+	core::dimension2du size_top = top->getDimension();
+	core::dimension2du size_left = left->getDimension();
+	core::dimension2du size_right = right->getDimension();
+
+	// Calculate image size to match textures size, but make sure it is
+	// 64px at least as it would look bad otherwise.
+	int size = npot2(std::max({64u,
+		size_top.Width, size_top.Height,
+		size_left.Width, size_left.Height,
+		size_right.Width, size_right.Height,
+	}));
 
-		guiScalingCache(io::path(params.rtt_texture_name.c_str()), driver, inventory_image);
-
-		video::ITexture *rtt = driver->addTexture(params.rtt_texture_name.c_str(), inventory_image);
-		inventory_image->drop();
-
-		if (rtt == NULL) {
-			errorstream << "TextureSource::generateTextureFromMesh(): failed to recreate texture from image: " << params.rtt_texture_name << std::endl;
-			return NULL;
-		}
-
-		driver->makeColorKeyTexture(rtt, v2s32(0,0));
-
-		if (params.delete_texture_on_shutdown)
-			m_texture_trash.push_back(rtt);
-
-		return rtt;
-	}
-#endif
-
-	if (!driver->queryFeature(video::EVDF_RENDER_TO_TARGET)) {
-		static bool warned = false;
-		if (!warned)
-		{
-			errorstream<<"TextureSource::generateTextureFromMesh(): "
-				<<"EVDF_RENDER_TO_TARGET not supported."<<std::endl;
-			warned = true;
+	video::IVideoDriver *driver = RenderingEngine::get_video_driver();
+	video::IImage *result = driver->createImage(video::ECF_A8R8G8B8, {size, size});
+	result->fill(video::SColor(0x00000000u));
+
+	// Offset of the view space origin (from the top image edge),
+	// where the upper near cube vertex is. Full image is 2 units.
+	static constexpr float base_y = 0.9f;
+
+	// Cube dimensions. The image is 2x2 in this (view) space.
+	static constexpr float diag_x = 0.8f;
+	static constexpr float diag_y = 0.4f;
+	static constexpr float vert_y = 1.2f;
+
+	static constexpr float coef_x = 0.5f / diag_x;
+	static constexpr float coef_y = -0.5f / diag_y;
+	static constexpr float coef_z = -1 / vert_y;
+
+	float scale = 2.f / size;
+	video::SColor *pixel = reinterpret_cast<video::SColor *>(result->lock());
```

Comment:
is the internal storage format of `SColor` guaranteed to match `ECF_A8R8G8B8`?

Reply:
this looks like a bad idea to me

Reply:
SColor.h:199:
> This class must consist of only one u32 and must not use virtual functions.

SColor.h:446:
```c
//! color in A8R8G8B8 Format
u32 color;
```

That is, it is guaranteed to be 32-bit, consisting of single u32, which stores color in A8R8G8B8 format.